commons/src/main/java/org/infinispan/multimap/api/MultimapCache.java

```diff
@@ -0,0 +1,91 @@
+package org.infinispan.multimap.api;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Predicate;
+
+/**
+ * MultimapCache provides the common building block for the two different types of multimap caches that Infinispan
+ * provides: embedded and remote.
+ * <p>
+ * Please see the <a href="http://infinispan.org/documentation/">Infinispan documentation</a> and/or the <a
+ * href="http://infinispan.org/docs/dev/getting_started/getting_started.html">5 Minute
+ * Usage Tutorial</a> for more details.
+ * <p/>
+ *
+ * @author Katia Aresti, karesti@redhat.com
+ * @see <a href="http://infinispan.org/documentation/">Infinispan documentation</a>
+ * @since 9.1
+ */
+public interface MultimapCache<K, V> {
+
+   /**
+    * Puts a value on a key.
+    * If the multimap cache supports duplicates, the value will be always added.
+    * If the multimap cache does not support duplicates and the value exists on the key, nothing will be done.
+    *
+    * @param key, the key to be put
+    * @param value, the value to added
+    */
+   CompletableFuture<Void> put(K key, V value);
+
+   /***
+    * Returns a view collection of the values associated with key in this multimap cache, if any.
+    * Any changes to the retrieved collection won't change the values in this multimap cache.
+    *
+    * @param key, to be retrieved
+    * @return the collection found in the key when it's present,
+    *         empty Collection when the key is not present
+    */
+   CompletableFuture<Collection<V>> get(K key);
```

Comment:
This will be a copy. Collections.unmodifiableCollection(v)
Changes are not allowed, and if the user needs to modify the collection directly, it should create a new one from the answer.
What do you think ?